[train] Fix pause and resume logic and generation tests after 0.16.0 upgrade

# Summary

#1240 updated vllm to 0.16.0 but this breaks the current pause and resume logic and some generation tests [Internal link to failing job](https://console.anyscale.com/cld_hxkifz7xa22mwicp21nzkds1lw/prj_4b6c498rypyq6g7yhk6vzgjevt/jobs/prodjob_xnqaavvy8xy9x3edlsxrqxsjwg?job-logs-section-tabs=application_logs&job-tab=overview) . 

```bash
FAILED tests/backends/skyrl_train/gpu/gpu_ci/test_inference_engine_client_http_endpoint.py::test_http_endpoint_completions_routing_and_batching - ModuleNotFoundError: No module named 'vllm.entrypoints.openai.protocol'
FAILED tests/backends/skyrl_train/gpu/gpu_ci/test_inference_engine_client_http_endpoint.py::test_http_endpoint_openai_api_with_weight_sync - ModuleNotFoundError: No module named 'vllm.entrypoints.openai.protocol'
FAILED tests/backends/skyrl_train/gpu/gpu_ci/test_inference_engine_client_http_endpoint.py::test_context_length_error_returns_400 - AssertionError: Error message should mention 'maximum context length': You passed 1025 input tokens and requested 0 output tokens. However, the model's context length is only 1024 tokens, resulting in a maximum input length of 1024 tokens. Please reduce the length of the input prompt. (parameter=input_tokens, value=1025)
assert 'maximum context length' in "you passed 1025 input tokens and requested 0 output tokens. however, the model's context length is only 1024 tokens, ...aximum input length of 1024 tokens. please reduce the length of the input prompt. (parameter=input_tokens, value=1025)"
 +  where "you passed 1025 input tokens and requested 0 output tokens. however, the model's context length is only 1024 tokens, ...aximum input length of 1024 tokens. please reduce the length of the input prompt. (parameter=input_tokens, value=1025)" = <built-in method lower of str object at 0x7a37507fb930>()
 +    where <built-in method lower of str object at 0x7a37507fb930> = "You passed 1025 input tokens and requested 0 output tokens. However, the model's context length is only 1024 tokens, ...aximum input length of 1024 tokens. Please reduce the length of the input prompt. (parameter=input_tokens, value=1025)".lower
FAILED tests/backends/skyrl_train/gpu/gpu_ci/test_pause_and_continue_generation.py::test_abort_generation_vllm_engine - AssertionError: assert 'length' == 'abort'
  
  - abort
  + length
```

Note that these are all related to the default inference codepath (with `InferenceEngineClient` + `RayWrappedInferenceEngine`). A related ticket is a similar list of issues with the new inference servers codepath #1242

--- a/skyrl/backends/skyrl_train/inference_engines/vllm/vllm_engine.py
+++ b/skyrl/backends/skyrl_train/inference_engines/vllm/vllm_engine.py
@@ -170,6 +170,18 @@ class BaseVLLMInferenceEngine(InferenceEngineInterface):
         """Get the underlying engine for RPC calls."""
         return self.llm.engine if hasattr(self.llm, "engine") else self.llm
 
+    @staticmethod
+    def _get_unfinished_request_ids(output_processor) -> list:
+        """Get unfinished request IDs suitable for abort/abort_request calls.
+
+        In vllm 0.16.0+, request_states is keyed by internal IDs (with a random suffix),
+        while abort() expects external IDs by default. We use external_req_ids when
+        available and fall back to request_states keys for older vllm versions.
+        """
+        if hasattr(output_processor, "external_req_ids"):
+            return list(output_processor.external_req_ids.keys())
+        return list(output_processor.request_states.keys())
+
     def reset_prefix_cache(self):
         """Reset the prefix cache. Subclasses override for async version."""
         return self.llm.llm_engine.reset_prefix_cache()
@@ -245,7 +257,7 @@ class VLLMInferenceEngine(BaseVLLMInferenceEngine):
                 "generation should be done before sleep() is called. Check for potential failures or "
                 "dangling requests in your Generator/Env. Aborting all unfinished requests."
             )
-            unfinished_request_ids = list(output_processor.request_states.keys())
+            unfinished_request_ids = self._get_unfinished_request_ids(output_processor)
             await asyncio.to_thread(engine.abort_request, unfinished_request_ids)
 
         level = 1 if self._is_lora else kwargs.get("level", 2)
@@ -457,7 +469,7 @@ class AsyncVLLMInferenceEngine(BaseVLLMInferenceEngine):
                 "generation should be done before sleep() is called. Check for potential failures or "
                 "dangling requests in your Generator/Env. Aborting all unfinished requests."
             )
-            unfinished_request_ids = list(output_processor.request_states.keys())
+            unfinished_request_ids = self._get_unfinished_request_ids(output_processor)
             await engine.abort(unfinished_request_ids)
 
         # TODO(team): remove once vllm fixes this
@@ -608,8 +620,7 @@ class AsyncVLLMInferenceEngine(BaseVLLMInferenceEngine):
         already-generated tokens with a stop_reason of "abort".
         """
         engine = self._get_engine()
-        # Collect all request IDs currently tracked by the scheduler/output processor
-        unfinished_request_ids = list(engine.output_processor.request_states.keys())
+        unfinished_request_ids = self._get_unfinished_request_ids(engine.output_processor)
         if unfinished_request_ids:
             await engine.abort(unfinished_request_ids)
         await engine.reset_prefix_cache()  # avoid KV-cache pollution

